{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a54bb2f2-08c4-4040-8906-1191164b273b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ðŸ”„ Processing sample: 10KP_RG\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "07:11:47.468 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/usr/bin/gatk-4.6.2.0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "07:11:47.673 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
      "07:11:47.676 INFO  HaplotypeCaller - The Genome Analysis Toolkit (GATK) v4.6.2.0\n",
      "07:11:47.676 INFO  HaplotypeCaller - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
      "07:11:47.676 INFO  HaplotypeCaller - Executing as chdahl@genomics1 on Linux v5.4.0-214-generic amd64\n",
      "07:11:47.676 INFO  HaplotypeCaller - Java runtime: OpenJDK 64-Bit Server VM v17.0.15+6-Ubuntu-0ubuntu120.04\n",
      "07:11:47.676 INFO  HaplotypeCaller - Start Date/Time: May 7, 2025 at 7:11:47 AM UTC\n",
      "07:11:47.677 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
      "07:11:47.677 INFO  HaplotypeCaller - ------------------------------------------------------------\n",
      "07:11:47.677 INFO  HaplotypeCaller - HTSJDK Version: 4.2.0\n",
      "07:11:47.678 INFO  HaplotypeCaller - Picard Version: 3.4.0\n",
      "07:11:47.678 INFO  HaplotypeCaller - Built for Spark Version: 3.5.0\n",
      "07:11:47.680 INFO  HaplotypeCaller - HTSJDK Defaults.COMPRESSION_LEVEL : 2\n",
      "07:11:47.680 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false\n",
      "07:11:47.680 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true\n",
      "07:11:47.680 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false\n",
      "07:11:47.681 INFO  HaplotypeCaller - Deflater: IntelDeflater\n",
      "07:11:47.681 INFO  HaplotypeCaller - Inflater: IntelInflater\n",
      "07:11:47.681 INFO  HaplotypeCaller - GCS max retries/reopens: 20\n",
      "07:11:47.681 INFO  HaplotypeCaller - Requester pays: disabled\n",
      "07:11:47.681 INFO  HaplotypeCaller - Initializing engine\n",
      "07:11:47.846 INFO  HaplotypeCaller - Done initializing engine\n",
      "07:11:47.862 INFO  NativeLibraryLoader - Loading libgkl_utils.so from jar:file:/usr/bin/gatk-4.6.2.0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_utils.so\n",
      "07:11:47.864 INFO  NativeLibraryLoader - Loading libgkl_smithwaterman.so from jar:file:/usr/bin/gatk-4.6.2.0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_smithwaterman.so\n",
      "07:11:47.865 INFO  SmithWatermanAligner - Using AVX accelerated SmithWaterman implementation\n",
      "07:11:47.867 INFO  HaplotypeCallerEngine - Disabling physical phasing, which is supported only for reference-model confidence output\n",
      "07:11:47.877 INFO  NativeLibraryLoader - Loading libgkl_pairhmm_omp.so from jar:file:/usr/bin/gatk-4.6.2.0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_pairhmm_omp.so\n",
      "07:11:47.887 INFO  IntelPairHmm - Flush-to-zero (FTZ) is enabled when running PairHMM\n",
      "07:11:47.887 INFO  IntelPairHmm - Available threads: 60\n",
      "07:11:47.887 INFO  IntelPairHmm - Requested threads: 30\n",
      "07:11:47.887 INFO  PairHMM - Using the OpenMP multi-threaded AVX-accelerated native PairHMM implementation\n",
      "07:11:47.891 INFO  VectorLoglessPairHMM - Time spent in setup for JNI call : 0.0\n",
      "07:11:47.891 INFO  PairHMM - Total compute time in PairHMM computeLogLikelihoods() : 0.0\n",
      "07:11:47.891 INFO  SmithWatermanAligner - Total compute time in native Smith-Waterman : 0.00 sec\n",
      "07:11:47.892 INFO  HaplotypeCaller - Shutting down engine\n",
      "[May 7, 2025 at 7:11:47 AM UTC] org.broadinstitute.hellbender.tools.walkers.haplotypecaller.HaplotypeCaller done. Elapsed time: 0.01 minutes.\n",
      "Runtime.totalMemory()=285212672\n",
      "htsjdk.samtools.util.RuntimeIOException: File not found: /mnt/jupiter/johnsonlab/Capstone_proj/results/BQSR/10KP_RG.raw.vcf\n",
      "\tat htsjdk.variant.variantcontext.writer.VariantContextWriterBuilder.build(VariantContextWriterBuilder.java:451)\n",
      "\tat htsjdk.variant.variantcontext.writer.VariantContextWriterBuilder.build(VariantContextWriterBuilder.java:415)\n",
      "\tat org.broadinstitute.hellbender.utils.variant.GATKVariantContextUtils.createVCFWriter(GATKVariantContextUtils.java:128)\n",
      "\tat org.broadinstitute.hellbender.tools.walkers.haplotypecaller.HaplotypeCallerEngine.makeVCFWriter(HaplotypeCallerEngine.java:502)\n",
      "\tat org.broadinstitute.hellbender.tools.walkers.haplotypecaller.HaplotypeCaller.onTraversalStart(HaplotypeCaller.java:270)\n",
      "\tat org.broadinstitute.hellbender.engine.GATKTool.doWork(GATKTool.java:1117)\n",
      "\tat org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:150)\n",
      "\tat org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:203)\n",
      "\tat org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:222)\n",
      "\tat org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:166)\n",
      "\tat org.broadinstitute.hellbender.Main.mainEntry(Main.java:209)\n",
      "\tat org.broadinstitute.hellbender.Main.main(Main.java:306)\n",
      "Caused by: java.nio.file.AccessDeniedException: /mnt/jupiter/johnsonlab/Capstone_proj/results/BQSR/10KP_RG.raw.vcf\n",
      "\tat java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:90)\n",
      "\tat java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)\n",
      "\tat java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)\n",
      "\tat java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:218)\n",
      "\tat java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:484)\n",
      "\tat java.base/java.nio.file.Files.newOutputStream(Files.java:228)\n",
      "\tat htsjdk.variant.variantcontext.writer.VariantContextWriterBuilder.build(VariantContextWriterBuilder.java:447)\n",
      "\t... 11 more\n",
      "Using GATK jar /usr/bin/gatk-4.6.2.0/gatk-package-4.6.2.0-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /usr/bin/gatk-4.6.2.0/gatk-package-4.6.2.0-local.jar HaplotypeCaller -R /mnt/jupiter/johnsonlab/Capstone_proj/annotation_jasmine-the2899-mb-hirise-ahy4t__02-20-2022__hic_output.fasta -I /mnt/jupiter/johnsonlab/Capstone_proj/results/BAM/Sorted_RG/10KP_RG.sorted.bam -O /mnt/jupiter/johnsonlab/Capstone_proj/results/BQSR/10KP_RG.raw.vcf --native-pair-hmm-threads 30\n",
      "07:11:49.600 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/usr/bin/gatk-4.6.2.0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "07:11:49.703 INFO  VariantFiltration - ------------------------------------------------------------\n",
      "07:11:49.706 INFO  VariantFiltration - The Genome Analysis Toolkit (GATK) v4.6.2.0\n",
      "07:11:49.706 INFO  VariantFiltration - For support and documentation go to https://software.broadinstitute.org/gatk/\n",
      "07:11:49.706 INFO  VariantFiltration - Executing as chdahl@genomics1 on Linux v5.4.0-214-generic amd64\n",
      "07:11:49.706 INFO  VariantFiltration - Java runtime: OpenJDK 64-Bit Server VM v17.0.15+6-Ubuntu-0ubuntu120.04\n",
      "07:11:49.706 INFO  VariantFiltration - Start Date/Time: May 7, 2025 at 7:11:49 AM UTC\n",
      "07:11:49.706 INFO  VariantFiltration - ------------------------------------------------------------\n",
      "07:11:49.706 INFO  VariantFiltration - ------------------------------------------------------------\n",
      "07:11:49.708 INFO  VariantFiltration - HTSJDK Version: 4.2.0\n",
      "07:11:49.708 INFO  VariantFiltration - Picard Version: 3.4.0\n",
      "07:11:49.708 INFO  VariantFiltration - Built for Spark Version: 3.5.0\n",
      "07:11:49.710 INFO  VariantFiltration - HTSJDK Defaults.COMPRESSION_LEVEL : 2\n",
      "07:11:49.710 INFO  VariantFiltration - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false\n",
      "07:11:49.710 INFO  VariantFiltration - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true\n",
      "07:11:49.710 INFO  VariantFiltration - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false\n",
      "07:11:49.710 INFO  VariantFiltration - Deflater: IntelDeflater\n",
      "07:11:49.711 INFO  VariantFiltration - Inflater: IntelInflater\n",
      "07:11:49.711 INFO  VariantFiltration - GCS max retries/reopens: 20\n",
      "07:11:49.711 INFO  VariantFiltration - Requester pays: disabled\n",
      "07:11:49.711 INFO  VariantFiltration - Initializing engine\n",
      "Using GATK jar /usr/bin/gatk-4.6.2.0/gatk-package-4.6.2.0-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /usr/bin/gatk-4.6.2.0/gatk-package-4.6.2.0-local.jar VariantFiltration -V /mnt/jupiter/johnsonlab/Capstone_proj/results/BQSR/10KP_RG.raw.vcf --filter-expression QD < 2.0 || FS > 60.0 || MQ < 40.0 --filter-name FAIL -O /mnt/jupiter/johnsonlab/Capstone_proj/results/BQSR/10KP_RG.filtered.vcf\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 45\u001b[0m\n\u001b[1;32m     36\u001b[0m subprocess\u001b[38;5;241m.\u001b[39mrun([\n\u001b[1;32m     37\u001b[0m     gatk_path, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mHaplotypeCaller\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     38\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-R\u001b[39m\u001b[38;5;124m\"\u001b[39m, reference,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     41\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m--native-pair-hmm-threads\u001b[39m\u001b[38;5;124m\"\u001b[39m, threads\n\u001b[1;32m     42\u001b[0m ])\n\u001b[1;32m     44\u001b[0m \u001b[38;5;66;03m# === Step 2: Hard Filtering ===\u001b[39;00m\n\u001b[0;32m---> 45\u001b[0m \u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\n\u001b[1;32m     46\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgatk_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mVariantFiltration\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     47\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m-V\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mraw_vcf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     48\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m--filter-expression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mQD < 2.0 || FS > 60.0 || MQ < 40.0\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     49\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m--filter-name\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mFAIL\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     50\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m-O\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfiltered_vcf\u001b[49m\n\u001b[1;32m     51\u001b[0m \u001b[43m\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     52\u001b[0m subprocess\u001b[38;5;241m.\u001b[39mrun([\n\u001b[1;32m     53\u001b[0m     gatk_path, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSelectVariants\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     54\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-V\u001b[39m\u001b[38;5;124m\"\u001b[39m, filtered_vcf,\n\u001b[1;32m     55\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m--exclude-filtered\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     56\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-O\u001b[39m\u001b[38;5;124m\"\u001b[39m, passonly_vcf\n\u001b[1;32m     57\u001b[0m ])\n\u001b[1;32m     59\u001b[0m \u001b[38;5;66;03m# === Step 3: BQSR ===\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3.8/subprocess.py:495\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    493\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Popen(\u001b[38;5;241m*\u001b[39mpopenargs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs) \u001b[38;5;28;01mas\u001b[39;00m process:\n\u001b[1;32m    494\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 495\u001b[0m         stdout, stderr \u001b[38;5;241m=\u001b[39m \u001b[43mprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcommunicate\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    496\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m TimeoutExpired \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m    497\u001b[0m         process\u001b[38;5;241m.\u001b[39mkill()\n",
      "File \u001b[0;32m/usr/lib/python3.8/subprocess.py:1020\u001b[0m, in \u001b[0;36mPopen.communicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1018\u001b[0m         stderr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mread()\n\u001b[1;32m   1019\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mclose()\n\u001b[0;32m-> 1020\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1021\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1022\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/lib/python3.8/subprocess.py:1083\u001b[0m, in \u001b[0;36mPopen.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1081\u001b[0m     endtime \u001b[38;5;241m=\u001b[39m _time() \u001b[38;5;241m+\u001b[39m timeout\n\u001b[1;32m   1082\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1083\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1084\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1085\u001b[0m     \u001b[38;5;66;03m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[1;32m   1086\u001b[0m     \u001b[38;5;66;03m# The first keyboard interrupt waits briefly for the child to\u001b[39;00m\n\u001b[1;32m   1087\u001b[0m     \u001b[38;5;66;03m# exit under the common assumption that it also received the ^C\u001b[39;00m\n\u001b[1;32m   1088\u001b[0m     \u001b[38;5;66;03m# generated SIGINT and will exit rapidly.\u001b[39;00m\n\u001b[1;32m   1089\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/lib/python3.8/subprocess.py:1806\u001b[0m, in \u001b[0;36mPopen._wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1804\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1805\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m  \u001b[38;5;66;03m# Another thread waited.\u001b[39;00m\n\u001b[0;32m-> 1806\u001b[0m (pid, sts) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_try_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1807\u001b[0m \u001b[38;5;66;03m# Check the pid and loop as waitpid has been known to\u001b[39;00m\n\u001b[1;32m   1808\u001b[0m \u001b[38;5;66;03m# return 0 even without WNOHANG in odd situations.\u001b[39;00m\n\u001b[1;32m   1809\u001b[0m \u001b[38;5;66;03m# http://bugs.python.org/issue14396.\u001b[39;00m\n\u001b[1;32m   1810\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m pid \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid:\n",
      "File \u001b[0;32m/usr/lib/python3.8/subprocess.py:1764\u001b[0m, in \u001b[0;36mPopen._try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   1762\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[39;00m\n\u001b[1;32m   1763\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1764\u001b[0m     (pid, sts) \u001b[38;5;241m=\u001b[39m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwaitpid\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpid\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwait_flags\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1765\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mChildProcessError\u001b[39;00m:\n\u001b[1;32m   1766\u001b[0m     \u001b[38;5;66;03m# This happens if SIGCLD is set to be ignored or waiting\u001b[39;00m\n\u001b[1;32m   1767\u001b[0m     \u001b[38;5;66;03m# for child processes has otherwise been disabled for our\u001b[39;00m\n\u001b[1;32m   1768\u001b[0m     \u001b[38;5;66;03m# process.  This child is dead, we can't get the status.\u001b[39;00m\n\u001b[1;32m   1769\u001b[0m     pid \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import os\n",
    "\n",
    "reference = \"/mnt/jupiter/johnsonlab/Capstone_proj/annotation_jasmine-the2899-mb-hirise-ahy4t__02-20-2022__hic_output.fasta\"\n",
    "out_dir = \"/mnt/jupiter/johnsonlab/Capstone_proj/results/BQSR\"\n",
    "bam_dir = \"/mnt/jupiter/johnsonlab/Capstone_proj/results/BAM/Sorted_RG\"\n",
    "genome_file = \"/mnt/jupiter/johnsonlab/Capstone_proj/genome.txt\"\n",
    "gatk_path = \"/bin/gatk-4.6.2.0/gatk\"\n",
    "threads = \"30\"\n",
    "\n",
    "os.makedirs(out_dir, exist_ok=True)\n",
    "\n",
    "files = [\"10KP_RG\"]\n",
    "#\"14KP_RG\", \"15AZ_RG\", \"28KP_RG\", \"31SMW_RG\", \"32SMW_RG\", \"34SMW_RG\", \"35SMW_RG\", \"36SMW_RG\", \"38SMW_RG\", \"39SMW_RG\",\n",
    "#\"46SMW_RG\", \"52SMW_RG\", \"55SMW_RG\", \"57SMW_RG\", \"69SMF_RG\", \"70SMF_RG\", \"8KP_RG\", \"9KP_RG\", \"MOR021_RG\", \"MOR023_RG\"\n",
    "\n",
    "for file in files:\n",
    "    print(f\"ðŸ”„ Processing sample: {file}\")\n",
    "\n",
    "    sorted_rg = f\"{bam_dir}/{file}.sorted.bam\"\n",
    "    raw_vcf = f\"{out_dir}/{file}.raw.vcf\"\n",
    "    filtered_vcf = f\"{out_dir}/{file}.filtered.vcf\"\n",
    "    passonly_vcf = f\"{out_dir}/{file}.passonly.vcf\"\n",
    "    recal_table = f\"{out_dir}/{file}.recal.table\"\n",
    "    recal_bam = f\"{out_dir}/{file}.recalibrated.bam\"\n",
    "    final_gvcf = f\"{out_dir}/{file}.final.g.vcf\"\n",
    "    joint_vcf = f\"{out_dir}/{file}.joint_genotyped.vcf\"\n",
    "    snps_vcf = f\"{out_dir}/{file}.joint_snps.vcf\"\n",
    "    maf_filtered_vcf = f\"{out_dir}/{file}.selected_snps.vcf\"\n",
    "    snp_bed = f\"{out_dir}/{file}.snps.bed\"\n",
    "    snp_windows = f\"{out_dir}/{file}.snp_windows_30bp.bed\"\n",
    "    snp_counts = f\"{out_dir}/{file}.snps_with_counts.txt\"\n",
    "    final_snps_bed = f\"{out_dir}/{file}.final_snps.bed\"\n",
    "\n",
    "    # === Step 1: Initial Variant Calling ===\n",
    "    subprocess.run([\n",
    "        gatk_path, \"HaplotypeCaller\",\n",
    "        \"-R\", reference,\n",
    "        \"-I\", sorted_rg,\n",
    "        \"-O\", raw_vcf,\n",
    "        \"--native-pair-hmm-threads\", threads\n",
    "    ])\n",
    "\n",
    "    # === Step 2: Hard Filtering ===\n",
    "    subprocess.run([\n",
    "        gatk_path, \"VariantFiltration\",\n",
    "        \"-V\", raw_vcf,\n",
    "        \"--filter-expression\", \"QD < 2.0 || FS > 60.0 || MQ < 40.0\",\n",
    "        \"--filter-name\", \"FAIL\",\n",
    "        \"-O\", filtered_vcf\n",
    "    ])\n",
    "    subprocess.run([\n",
    "        gatk_path, \"SelectVariants\",\n",
    "        \"-V\", filtered_vcf,\n",
    "        \"--exclude-filtered\",\n",
    "        \"-O\", passonly_vcf\n",
    "    ])\n",
    "\n",
    "    # === Step 3: BQSR ===\n",
    "    subprocess.run([\n",
    "        gatk_path, \"BaseRecalibrator\",\n",
    "        \"-I\", sorted_rg,\n",
    "        \"-R\", reference,\n",
    "        \"--known-sites\", passonly_vcf,\n",
    "        \"-O\", recal_table,\n",
    "        \"--num-threads\", threads\n",
    "    ])\n",
    "    subprocess.run([\n",
    "        gatk_path, \"ApplyBQSR\",\n",
    "        \"-R\", reference,\n",
    "        \"-I\", sorted_rg,\n",
    "        \"--bqsr-recal-file\", recal_table,\n",
    "        \"-O\", recal_bam,\n",
    "        \"--num-threads\", threads\n",
    "    ])\n",
    "\n",
    "    # === Step 4: Final Variant Calling ===\n",
    "    subprocess.run([\n",
    "        gatk_path, \"HaplotypeCaller\",\n",
    "        \"-R\", reference,\n",
    "        \"-I\", recal_bam,\n",
    "        \"-O\", final_gvcf,\n",
    "        \"-ERC\", \"GVCF\",\n",
    "        \"--native-pair-hmm-threads\", threads\n",
    "    ])\n",
    "\n",
    "    # === Step 5: Joint Genotyping (single-sample placeholder) ===\n",
    "    subprocess.run([\n",
    "        gatk_path, \"GenotypeGVCFs\",\n",
    "        \"-R\", reference,\n",
    "        \"-V\", final_gvcf,\n",
    "        \"-O\", joint_vcf\n",
    "    ])\n",
    "\n",
    "    # === Step 6: Select SNPs only ===\n",
    "    subprocess.run([\n",
    "        gatk_path, \"SelectVariants\",\n",
    "        \"-V\", joint_vcf,\n",
    "        \"--select-type-to-include\", \"SNP\",\n",
    "        \"-O\", snps_vcf\n",
    "    ])\n",
    "\n",
    "    '''# === Step 7: MAF + missing data filtering ===\n",
    "    subprocess.run([\n",
    "        \"bcftools\", \"view\",\n",
    "        \"-i\", \"MAF>0.1 && F_MISSING<0.05\",\n",
    "        \"-o\", maf_filtered_vcf,\n",
    "        snps_vcf\n",
    "        #this doesn't work because there are no MAFs or F_MISSING added to the vcfs yet\n",
    "        #MAF: Minor Allele Frequency\n",
    "        #F_MISSING: Fraction of Missing Genotypes\n",
    "    ])'''\n",
    "    # === Step 7: Annotate with MAF and F_MISSING, then filter ===\n",
    "    annotated_vcf = f\"{out_dir}/{file}.annotated_snps.vcf\"\n",
    "\n",
    "    # Step 7a: Annotate SNPs VCF with MAF and F_MISSING\n",
    "    with open(annotated_vcf, 'w') as annot_out:\n",
    "        fill_tags = subprocess.Popen([\n",
    "            \"bcftools\", \"+fill-tags\", snps_vcf, \"-Ou\", \"--\", \"-t\", \"MAF,F_MISSING\"\n",
    "        ], stdout=subprocess.PIPE)\n",
    "        subprocess.run([\n",
    "            \"bcftools\", \"view\", \"-Ov\"\n",
    "        ], stdin=fill_tags.stdout, stdout=annot_out)\n",
    "\n",
    "    # Step 7b: Filter based on MAF > 0.1 and F_MISSING < 0.05\n",
    "    with open(maf_filtered_vcf, 'w') as filt_out:\n",
    "        subprocess.run([\n",
    "            \"bcftools\", \"view\",\n",
    "            \"-i\", \"MAF>0.1 && F_MISSING<0.05\",\n",
    "            annotated_vcf\n",
    "        ], stdout=filt_out)\n",
    "\n",
    "    # === Step 8: Convert SNPs to BED ===\n",
    "    with open(snp_bed, 'w') as bed_out:\n",
    "        subprocess.run([\n",
    "            \"bcftools\", \"query\",\n",
    "            \"-f\", \"%CHROM\\t%POS0\\t%POS\\t%ID\\n\",\n",
    "            maf_filtered_vcf\n",
    "        ], stdout=bed_out)\n",
    "\n",
    "    # === Step 9: Create 30bp windows ===\n",
    "    with open(snp_windows, 'w') as out_win:\n",
    "        subprocess.run([\n",
    "            \"bedtools\", \"slop\",\n",
    "            \"-i\", snp_bed,\n",
    "            \"-g\", genome_file,\n",
    "            \"-b\", \"30\"\n",
    "        ], stdout=out_win)\n",
    "\n",
    "    # === Step 10: Count SNPs per window ===\n",
    "    with open(snp_counts, 'w') as out_count:\n",
    "        subprocess.run([\n",
    "            \"bedtools\", \"intersect\",\n",
    "            \"-a\", snp_windows,\n",
    "            \"-b\", snp_bed,\n",
    "            \"-c\"\n",
    "        ], stdout=out_count)\n",
    "\n",
    "    # === Step 11: Filter SNPs without nearby neighbors ===\n",
    "    with open(snp_counts) as infile, open(final_snps_bed, 'w') as outfile:\n",
    "        for line in infile:\n",
    "            fields = line.strip().split('\\t')\n",
    "            if int(fields[4]) == 1:\n",
    "                outfile.write('\\t'.join(fields[:3]) + '\\n')\n",
    "\n",
    "    print(f\"âœ… Finished processing sample: {file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d6ef157-0d15-476b-9654-3edf56550051",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PIPELINE\n",
    "sorted.bam \n",
    "   â†“\n",
    "First HaplotypeCaller â†’ raw.vcf \n",
    "   â†“\n",
    "Hard Filtering â†’ passonly.vcf \n",
    "   â†“\n",
    "BaseRecalibrator + ApplyBQSR â†’ recalibrated.bam \n",
    "   â†“\n",
    "Second HaplotypeCaller â†’ final.g.vcf \n",
    "   â†“\n",
    "CombineGVCFs + GenotypeGVCFs â†’ joint.vcf \n",
    "   â†“\n",
    "Filtering (MAF, call rate) â†’ selected_snps.vcf \n",
    "   â†“\n",
    "Design SNP array / GT-seq panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "561f3e74-b2c1-40bf-9328-258222e90b8d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
