{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a0842bd-667c-4e53-bb8c-58a400379306",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Variant Calling (Mainly Using GATK tools)\n",
    "import subprocess\n",
    "import os\n",
    "\n",
    "reference = \"/mnt/jupiter/johnsonlab/Capstone_proj/annotation_jasmine-the2899-mb-hirise-ahy4t__02-20-2022__hic_output.fasta\"\n",
    "out_dir = \"/mnt/jupiter/johnsonlab/Capstone_proj/results/BQSR\"\n",
    "bam_dir = \"/mnt/jupiter/johnsonlab/Capstone_proj/results/test_results/BAM/Sorted\"\n",
    "genome_file = \"/mnt/jupiter/johnsonlab/Capstone_proj/genome.txt\"\n",
    "gatk_path = \"/bin/gatk-4.6.2.0/gatk\"\n",
    "threads = \"10\"\n",
    "\n",
    "os.makedirs(out_dir, exist_ok=True)\n",
    "\n",
    "files = [\"10KP\", \"14KP\", \"15AZ\", \"28KP\", \"31SMW\", \"32SMW\", \"34SMW\", \"35SMW\", \"36SMW\", \"38SMW\", \"39SMW\",\n",
    "         \"46SMW\", \"52SMW\", \"55SMW\", \"57SMW\", \"69SMF\", \"70SMF\", \"8KP\", \"9KP\", \"MOR021\", \"MOR023\"\n",
    "        ]\n",
    "\n",
    "for file in files:\n",
    "    print(f\"Processing sample: {file}\")\n",
    "\n",
    "    sorted_rg = f\"{bam_dir}/{file}.sorted.bam\"\n",
    "    raw_vcf = f\"{out_dir}/{file}.raw.vcf\"\n",
    "    filtered_vcf = f\"{out_dir}/{file}.filtered.vcf\"\n",
    "    passonly_vcf = f\"{out_dir}/{file}.passonly.vcf\"\n",
    "    recal_table = f\"{out_dir}/{file}.recal.table\"\n",
    "    recal_bam = f\"{out_dir}/{file}.recalibrated.bam\"\n",
    "    final_gvcf = f\"{out_dir}/{file}.final.g.vcf\"\n",
    "    joint_vcf = f\"{out_dir}/{file}.joint_genotyped.vcf\"\n",
    "    snps_vcf = f\"{out_dir}/{file}.joint_snps.vcf\"\n",
    "    maf_filtered_vcf = f\"{out_dir}/{file}.selected_snps.vcf\"\n",
    "    snp_bed = f\"{out_dir}/{file}.snps.bed\"\n",
    "    snp_windows = f\"{out_dir}/{file}.snp_windows_30bp.bed\"\n",
    "    snp_counts = f\"{out_dir}/{file}.snps_with_counts.txt\"\n",
    "    final_snps_bed = f\"{out_dir}/{file}.final_snps.bed\"\n",
    "\n",
    "        # Index the sorted BAM\n",
    "    subprocess.run([\n",
    "        \"samtools\", \"index\", sorted_rg\n",
    "    ], check=True)\n",
    "\n",
    "    # Step 1: Initial Variant Calling\n",
    "    subprocess.run([\n",
    "        gatk_path, \"HaplotypeCaller\",\n",
    "        \"-R\", reference,\n",
    "        \"-I\", sorted_rg,\n",
    "        \"-O\", raw_vcf,\n",
    "        \"--native-pair-hmm-threads\", threads\n",
    "    ])\n",
    "\n",
    "    # Step 2: Hard Filtering\n",
    "    subprocess.run([\n",
    "        gatk_path, \"VariantFiltration\",\n",
    "        \"-V\", raw_vcf,\n",
    "        \"--filter-expression\", \"QD < 2.0 || FS > 60.0 || MQ < 40.0\",\n",
    "        \"--filter-name\", \"FAIL\",\n",
    "        \"-O\", filtered_vcf\n",
    "    ])\n",
    "    subprocess.run([\n",
    "        gatk_path, \"SelectVariants\",\n",
    "        \"-V\", filtered_vcf,\n",
    "        \"--exclude-filtered\",\n",
    "        \"-O\", passonly_vcf\n",
    "    ])\n",
    "    # Step 3: Recall Table\n",
    "    # Takes about 5 minutes per file\n",
    "    subprocess.run([\n",
    "        \"/bin/gatk-4.6.2.0/gatk\", \"BaseRecalibrator\",\n",
    "        \"-I\", sorted_rg,\n",
    "        \"-R\", reference,\n",
    "        \"--known-sites\", passonly_vcf,\n",
    "        \"-O\", recal_table,\n",
    "    ])\n",
    "    # Step 4: Apply BQSR\n",
    "    subprocess.run([\n",
    "        # takes about 7 minutes per file\n",
    "        gatk_path, \"ApplyBQSR\",\n",
    "        \"-R\", reference,\n",
    "        \"-I\", sorted_rg,\n",
    "        \"--bqsr-recal-file\", recal_table,\n",
    "        \"-O\", recal_bam,\n",
    "    ])\n",
    "    # Step 5: Final Variant Calling\n",
    "    subprocess.run([\n",
    "        gatk_path, \"--java-options\", \"-Xmx16g\", \"HaplotypeCaller\",\n",
    "        \"-R\", reference,\n",
    "        \"-I\", recal_bam,\n",
    "        \"-O\", final_gvcf,\n",
    "        \"-ERC\", \"GVCF\",\n",
    "        \"--native-pair-hmm-threads\", threads\n",
    "    ])\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
